python/gcp_client/hydrotools/gcp_client/gcp.py

```diff
@@ -166,126 +182,209 @@ def get_blob(self, blob_name) -> bytes:
             The data stored in the blob.
         
         """
+        # Setup anonymous client and retrieve blob data
         client = storage.Client.create_anonymous_client()
         bucket = client.bucket(self.bucket_name)
         return bucket.blob(blob_name).download_as_bytes(timeout=120)
 
-    def get_DataFrame(self, blob_name, **kwargs) -> pd.DataFrame:
-        """Retrieve a blob from the data service as a pandas.DataFrame.
+    def get_Dataset(
+        self, 
+        blob_name: str,
+        feature_id_filter: Union[npt.ArrayLike, bool] = True
+        ) -> xr.Dataset:
+        """Retrieve a blob from the data service as xarray.Dataset
 
         Parameters
         ----------
         blob_name : str, required
             Name of blob to retrieve.
+        feature_id_filter : bool or array-like, optional, default False
+            If True, filter data using default list of feature ids (USGS gaging locations).
+            Alternatively, limit data returned to feature ids in feature_id_filter 
+            list.
 
         Returns
         -------
-        df : pandas.DataFrame
+        ds : xarray.Dataset
             The data stored in the blob.
         
         """
-        bytes_string = self.get_blob(blob_name)
-        return NWM_bytes_to_DataFrame(bytes_string, **kwargs)
+        # Get raw bytes
+        raw_bytes = self.get_blob(blob_name)
+
+        # Create Dataset
+        ds = xr.load_dataset(
+            BytesIO(raw_bytes),
+            engine='h5netcdf',
+            mask_and_scale=True
+            )
 
-    def _make_blob_name(self, configuration, reference_time, valid_hour) -> str:
-        """Generate blob name for retrieval.
+        # Attempt to filter Dataset
+        if isinstance(feature_id_filter, Iterable):
+            try:
+                feature_id_filter = list(feature_id_filter)
+                return ds.sel(feature_id=feature_id_filter)
+            except:
+                warnings.warn("Invalid feature_id_filter")
+                return ds
+
+        # Return default filtered Dataset
+        if feature_id_filter == True:
+            feature_id_filter = list(self.crosswalk.index)
+            return ds.sel(feature_id=feature_id_filter)
+
+        # Return unfiltered Dataset
```

Comment:
Here's the dilemma. Most users are unlikely to request a single dataset. To retrieve an entire time series involves retrieving and deserializing 3 to 120 files. I'm afraid memory use will exceed RAM on many systems. From a software development perspective, I understand why this method should not filter by default. From an engineering perspective, I think not filtering will result in a complete failure of the method for most users. The sequence of conditionals could certainly be cleaned up, but I think filtering by default is the safest option and the option that's most compatible with our principle use-case (data scientists).

Reply:
> The sequence of conditionals could certainly be cleaned up, but I think filtering by default is the safest option and the option that's most compatible with our principle use-case (data scientists).

Given your added context, I completely agree -- I actually think this was the default behavior in the previous implementation. Determining when to start writing to disk is definitely on the docket of task to tackle. It may be appropriate to move it up in importance.